agent/api/container/status/managedagentstatus.go

```diff
@@ -0,0 +1,112 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package status
+
+import (
+	"errors"
+	"strings"
+)
+
+const (
+	// ManagedAgentStatusNone is the zero state of a managed agent
+	ManagedAgentStatusNone ManagedAgentStatus = iota
+	// ManagedAgentCreated represents a managed agent which has dependencies in place
+	ManagedAgentCreated
+	// ManagedAgentRunning represents a managed agent that has started sucessfully
+	ManagedAgentRunning
+	// ManagedAgentStopped represents a managed agent that has stopped
+	ManagedAgentStopped
+)
+
+// ManagedAgentStatus is an enumeration of valid states in the managed agent lifecycle
+type ManagedAgentStatus int32
+
+var managedAgentStatusMap = map[string]ManagedAgentStatus{
+	"NONE":    ManagedAgentStatusNone,
+	"CREATED": ManagedAgentCreated,
+	"RUNNING": ManagedAgentRunning,
+	"STOPPED": ManagedAgentStopped,
+}
+
+// String returns a human readable string representation of this object
+func (mas ManagedAgentStatus) String() string {
+	for k, v := range managedAgentStatusMap {
+		if v == mas {
+			return k
+		}
+	}
+	return "NONE"
+}
+
+// Terminal returns true if the managed agent status is STOPPED
+func (mas ManagedAgentStatus) Terminal() bool {
+	return mas == ManagedAgentStopped
+}
+
+// IsRunning returns true if the managed agent status is either RUNNING
+func (mas ManagedAgentStatus) IsRunning() bool {
+	return mas == ManagedAgentRunning
+}
+
+// BackendStatus maps the internal managedAgent status in the agent to that in the backend
+func (mas ManagedAgentStatus) BackendStatus() string {
+	switch mas {
+	case ManagedAgentStatusNone:
+		return mas.String()
+	case ManagedAgentCreated:
+		fallthrough
+	case ManagedAgentRunning:
+		return mas.String()
+	case ManagedAgentStopped:
+		return mas.String()
+	}
+	return "PENDING"
+}
+
+// ShouldReportToBackend returns true if the managedAgent status is recognized as a
+// valid state by ECS. Note that not all managedAgent statuses are recognized by ECS
+// or map to ECS states
+// For now we expect PENDING/RUNNING/STOPPED as valid states
+// we should only skip events that have ManagedAgentStatusNone
+func (mas *ManagedAgentStatus) ShouldReportToBackend() bool {
```

Comment:
any reason why the receiver for this is a pointer but for the previous methods is an struct? (i.e. `mas *ManagedAgentStatus` vs `mas ManagedAgentStatus`

Reply:
Same for method below.

Reply:
+1, this function should be non-pointer I think, UnmarshalJSON probably has to be pointer though in order to fulfil the interface.

Reply:
stole it from: https://github.com/aws/amazon-ecs-agent/blob/782948476da6d995ad33c6a04130f8172820af27/agent/api/container/status/containerstatus.go#L97-L99
I'll make the update and test.